### ISSUE TITLE
IllegalArgumentException: 'other' has different root on windows

### ISSUE THREAD

**Issue by OlegYch** [swoval/swoval#147]
every other sbt reload on windows throws
```
[error] java.lang.IllegalArgumentException: 'other' has different root
[error]         at sun.nio.fs.WindowsPath.relativize(Unknown Source)
[error]         at sun.nio.fs.WindowsPath.relativize(Unknown Source)
[error]         at com.swoval.files.CachedDirectoryImpl.update(CachedDirectoryImpl.java:245)
[error]         at com.swoval.files.NioPathWatcher.update(NioPathWatcher.java:354)
[error]         at com.swoval.files.NioPathWatcher.register(NioPathWatcher.java:201)
[error]         at com.swoval.files.FileCacheDirectoryTree.register(FileCacheDirectoryTree.java:417)
[error]         at com.swoval.files.FileCachePathWatcher.register(FileCachePathWatcher.java:24)
[error]         at com.swoval.files.FileTreeRepositoryImpl.register(FileTreeRepositoryImpl.java:101)
[error]         at sbt.internal.nio.FileTreeRepositoryImpl.register(FileTreeRepositoryImpl.scala:130)
[error]         at sbt.nio.Settings$.$anonfun$inputPathSettings$3(Settings.scala:157)
[error]         at sbt.nio.Settings$.$anonfun$inputPathSettings$3$adapted(Settings.scala:157)
[error]         at scala.collection.immutable.List.foreach(List.scala:431)
[error]         at sbt.nio.Settings$.$anonfun$inputPathSettings$2(Settings.scala:157)
[error]         at sbt.nio.Settings$.$anonfun$inputPathSettings$2$adapted(Settings.scala:156)
[error]         at scala.Option.foreach(Option.scala:407)
[error]         at sbt.nio.Settings$.$anonfun$inputPathSettings$1(Settings.scala:156)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error]         at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error]         at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error]         at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error]         at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error]         at sbt.Execute.work(Execute.scala:291)
[error]         at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error]         at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error]         at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error]         at java.util.concurrent.FutureTask.run(Unknown Source)
[error]         at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
[error]         at java.util.concurrent.FutureTask.run(Unknown Source)
[error]         at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
[error]         at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
[error]         at java.lang.Thread.run(Unknown Source)
[error] (Compile / unmanagedSources / allInputPathsAndAttributes) java.lang.IllegalArgumentException: 'other' has different root
```
https://github.com/sbt/sbt/issues/6471

**Comment by eed3si9n** [comment]
I am guessing that it happens when you try to relativize two Windows paths with different drive letters.
In this case, the global plugin subproject ` $HOME/.sbt/1.0/plugins` in `C:\` but the project in `E:\` etc.

**Comment by OlegYch** [comment]
@eed3si9n that was my guess as well, but nope

**Comment by eatkins** [comment]
@OlegYch thanks for the report. I cannot reproduce on my VM though so I'm going to need a better repro to be able to actually push a fix or you are going to need to publish your own local copy of swoval+sbt with extra debug logging so we can confirm what is happening.

The issue is happening here:
https://github.com/swoval/swoval/blob/44077d270c04d1b804405c2851166aab9b2ebdc7/files/jvm/src/main/java/com/swoval/files/CachedDirectoryImpl.java#L240-L248

Theoretically the pathFilter associated with the directory should reject a file that isn't in the same partition or the directory, so I don't really understand why there would be a problem with the root. The naive fix would be to check if `typedPath.getPath().startsWith(this.getPath())` and skip the call to updateImpl if it doesn't, but I don't want to make that change without understanding what is going on and confirming that it would actually fix your change. 

What would be helpful would be for you to clone swoval, add some printlns that show what `this.getPath()` and `typedPath.getPath()` return. Then you'd need to add something like `libraryDependencies += "com.swoval" % "file-tree-views" % "2.2.0-SNAPSHOT"` to sbt's build.sbt in the root project: https://github.com/sbt/sbt/blob/2a26e8186894dc0af46cd142306727b1acd722f8/build.sbt#L179 and publish a snapshot version of sbt and report back what is printed when you generate that error.

**Comment by OlegYch** [comment]
ok found the cause - i have ~/.sbt symlinked to another drive
so on every other reload `System.out.println(typedPath.getPath() + "->" + this.getPath());`
prints `C:\Users\OlegYch\.sbt\1.0\plugins->D:\Distrib\Coding\Java\.sbt\1.0` and throws

**Comment by OlegYch** [comment]
i guess it should not try to resolve the symlink

**Comment by eatkins** [comment]
@OlegYch thanks for the update. I was able to reproduce on my vm and track down the problem: #150.

**Comment by eatkins** [comment]
I released version 2.1.7 which should have the fix for this and opened a pr in sbt io https://github.com/sbt/io/pull/323.
